chronos/ratelimit.nim

```diff
@@ -0,0 +1,119 @@
+
+export timer
+
+type
+  BucketWaiter = object
+    future: Future[void]
+    value: int
+    alreadyConsumed: int
+
+  TokenBucket* = ref object
+    budget: int
+    budgetCap: int
+    lastUpdate: Moment
+    fillPerMs: int
+    workFuture: Future[void]
+    pendingRequests: seq[BucketWaiter]
+    manuallyReplenished: AsyncEvent
+
+proc update(bucket: TokenBucket) =
+  let
+    currentTime = Moment.now()
+    timeDelta = milliseconds(currentTime - bucket.lastUpdate).int
+    replenished = bucket.fillPerMs * timeDelta
+
+  bucket.lastUpdate += timeDelta.milliseconds
+  bucket.budget = min(bucket.budgetCap, bucket.budget + replenished)
+
+proc tryConsume*(bucket: TokenBucket, tokens: int): bool =
+  ## If `tokens` are available, consume them,
+  ## Otherwhise, return false.
+
+  bucket.update()
+
+  if bucket.budget >= tokens:
+    bucket.budget -= tokens
+    true
+  else:
+    false
+
+proc worker(bucket: TokenBucket) {.async.} =
+  while bucket.pendingRequests.len > 0:
+    bucket.manuallyReplenished.clear()
+    template waiter: untyped = bucket.pendingRequests[0]
+
+    if bucket.tryConsume(waiter.value):
+      waiter.future.complete()
+      bucket.pendingRequests.delete(0)
+    else:
+      waiter.value -= bucket.budget
+      waiter.alreadyConsumed.inc(bucket.budget)
+      bucket.budget = 0
+
+      let eventWaiter = bucket.manuallyReplenished.wait()
+      if bucket.fillPerMs > 0:
+        let
+          nextCycleValue = min(waiter.value, bucket.budgetCap)
+          timeToZero = (nextCycleValue div bucket.fillPerMs) + 1
+          sleeper = sleepAsync(milliseconds(timeToZero))
+        await sleeper or eventWaiter
+        sleeper.cancel()
+        eventWaiter.cancel()
+      else:
+        await eventWaiter
+
+  bucket.workFuture = nil
+
+proc consume*(bucket: TokenBucket, tokens: int): Future[void] =
+  ## Wait for `tokens` to be available, and consume them.
+
+  let retFuture = newFuture[void]("TokenBucket.consume")
+  if isNil(bucket.workFuture) or bucket.workFuture.finished():
+    if bucket.tryConsume(tokens):
+      retFuture.complete()
+      return retFuture
+
+  bucket.pendingRequests.add(BucketWaiter(future: retFuture, value: tokens))
+  if isNil(bucket.workFuture) or bucket.workFuture.finished():
+    bucket.workFuture = worker(bucket)
+
+  proc cancellation(udata: pointer) =
+    for index in 0..<bucket.pendingRequests.len:
+      if bucket.pendingRequests[index].future == retFuture:
+        bucket.budget.inc(bucket.pendingRequests[index].alreadyConsumed)
+        bucket.pendingRequests.delete(index)
+        if index == 0:
+          bucket.manuallyReplenished.fire()
+        break
+  retFuture.cancelCallback = cancellation
+  return retFuture
+
+proc replenish*(bucket: TokenBucket, tokens: int) =
+  ## Add `tokens` to the budget (capped to the bucket capacity)
+  bucket.budget += tokens
+  bucket.update()
+  bucket.manuallyReplenished.fire()
+
+proc new*(
+  T: type[TokenBucket],
+  budgetCap, fillPerMs: int): T =
```

Comment:
this fill rate should probably be per second (SI units), and as such, a float, in the public api at least

Reply:
If we expose a float and use a int per ms internally, it may be surprising to user since there will be rounding errors (ie, anything less than 1 / ms will be rounded to 0, so can't use 10 per second)
And can't remember why, but a design goal was to avoid floats internally

Reply:
So we either switch to float internally, (would need to remember why that wasn't done in the first place), or we switch to seconds internally, which would make the bucket update only once per second

Reply:
> rounded to 0

only as implemented presently

Reply:
Or `assert fillPerSecond mod 1000 == 0` :D

Reply:
It might be that some operating systems (like Linux) can (potentially) do faster context switches (no save & restore of FP context) until you first dirty floating point registers.  Once you do that the OS sets a flag and always does it.

This may be less relevant in the modern era of SIMD-optimized string libraries since almost all code does (string OR float) stuff.  OTOH, maybe on x86 it is only the FP stack (in|ex)cluded, and that might be a smaller % of state to save&restore anyway with all the (SSE/AVX)* registers.  So, it could also be less of an optimization these days.

It would be good to point to some document somewhere on the inet elaborating on this (and trying, hah!, to measure it).  Whether to commit to float in otherwise context switch heavy code like networking certainly comes up a lot.

Reply:
anyway, we could even stick with `int` and still use seconds in init - this is after all just a scaling factor.. internally the library can keep using `millis` or even `micros` for its computations as long as the "external" api is sane .. this is not really a real-time API - things like rounding and granularity are somewhat expected and as long as they're dealt with with reasonable attention to drift/bias, it seems somewhat arbitrary - so if I was using a fillPerMS of 1000 before, I'd use `1` now and if I want `0.5` I'd simply scale my entire token allocation by 10x and use `5`.

Reply:
alternative: turn this parameter into a `replenishTime: Duration`

Reply:
Switched to float internally, and all tests fail, ie
```
budget = 1000
tryConsume(800) == true
tryConsume(200) == false # rounding error
```
So I'll fix them by adding more margins, but still something to have in mind

> anyway, we could even stick with int and still use seconds in init

The issue is that currently, we can "tick" the bucket every millisecond, so we must be able to update its budget every millisecond. If you have fill rate per second = 1, what is the budget increase of 1ms supposed to be? 0.001, which requires a float budget

Which is why the two options in my eyes are: 1) use float internally 2) change the "tick" to second 3) keep fill in milliseconds
1. may indeed be the lesser evil, as long as we are careful not to add up errors over time

Reply:
Oh, maybe we could just scale everything to 1000* indeed

Reply:
Oh, maybe we could just scale everything to 1000* indeed

Reply:
`replenishTime` fixes rounding and allows arbitrary granularity internally (per-second granularity is probably not good enough ... millis or micros at least)

Reply:
Oh, github played games and hid your comment before I replied
Sounds good!

Reply:
Done, still requires some floatery, but it's at least contained & cannot drift